cmd/jujud/agent/model/manifolds.go

```diff
@@ -328,30 +328,30 @@ func IAASManifolds(config ManifoldsConfig) dependency.Manifolds {
 		// it.
 
 		// The undertaker is currently the only ifNotAlive worker.
-		undertakerName: ifNotUpgrading(ifNotAlive(undertaker.Manifold(undertaker.ManifoldConfig{
+		undertakerName: ifNotUpgrading(ifNotAlive(ifCredentialValid(undertaker.Manifold(undertaker.ManifoldConfig{
 			APICallerName:      apiCallerName,
 			CloudDestroyerName: environTrackerName,
 
 			NewFacade:                    undertaker.NewFacade,
 			NewWorker:                    undertaker.NewWorker,
 			NewCredentialValidatorFacade: common.NewCredentialInvalidatorFacade,
-		}))),
+		})))),
 
 		// All the rest depend on ifNotMigrating.
-		computeProvisionerName: ifNotMigrating(provisioner.Manifold(provisioner.ManifoldConfig{
+		computeProvisionerName: ifNotMigrating(ifCredentialValid(provisioner.Manifold(provisioner.ManifoldConfig{
 			AgentName:                    agentName,
 			APICallerName:                apiCallerName,
 			EnvironName:                  environTrackerName,
 			NewProvisionerFunc:           provisioner.NewEnvironProvisioner,
 			NewCredentialValidatorFacade: common.NewCredentialInvalidatorFacade,
-		})),
+		}))),
 		storageProvisionerName: ifNotMigrating(storageprovisioner.ModelManifold(storageprovisioner.ModelManifoldConfig{
```

Comment:
Doesn't the storage provisioner require a valid credential?
eg on AWS, it will talk to the cloud to create EBS volumes

Reply:
It does :) This is why this worker will be stopped if model's cloud credential is not valid since 'ifNotMigrating' is a flag worker that also depends on valid credential and it will not be running.

I will be adding the context with callback capable of invalidating credentials to all storage related cloud calls very soon. For now, storage workers will be stopped indirectly anyway :D